File: baekjoon_problems/exhaustive_search/14889.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.abs
import kotlin.math.min
import kotlin.system.exitProcess

private var minDiff = Int.MAX_VALUE

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val n = readLine().toInt()
    // 점수 2차원 그래프
    val graph = Array(n) { IntArray(n) }
    // 방문을 확인하기 위한 리스트
    val visited = BooleanArray(n)

    repeat(n) { row ->
        val st = StringTokenizer(readLine(), " ")
        repeat(n) { col ->
            graph[row][col] = st.nextToken().toInt()
        }
    }

    // 차이를 계산하는 함수
    fun diff() {
        var startTeam = 0
        var linkTeam = 0
        
        // 만들어진 팀에서 조합을 계산
        for(i in 0 until n - 1) {
            for(j in i + 1 until n) {
                // i번째 사람과 j번째 사람을 방문했다면 스타트팀으로 점수 계산
                if(visited[i] && visited[j]) {
                    startTeam += graph[i][j]
                    startTeam += graph[j][i]
                }
                // i번째 사람과 j번째 사람을 방문하지 않았다면 링크팀으로 점수 계산
                else if(!visited[i] && !visited[j]) {
                    linkTeam += graph[i][j]
                    linkTeam += graph[j][i]
                }
            }
        }

        // 두 팀의 차이 계산
        val diff = abs(startTeam - linkTeam)

        // 점수가 0점이라면 최소 차이이므로 바로 출력
        if(diff == 0) {
            println(diff)
            exitProcess(0)
        }
        
        // 최소 차이 갱신
        minDiff = min(diff, minDiff)
    }

    // DFS 함수
    fun dfs(index: Int, cnt: Int) {
        // 팀이 나누어졌다면
        if(cnt == n / 2) {
            diff()
            return
        }

        for(i in index until n) {
            // 현재 선수를 방문하지 않았다면
            if(!visited[i]) {
                // 방문 처리 후 DFS 호출
                visited[i] = true
                dfs(i + 1, cnt + 1)
                // 다음을 위해 방문 처리 취소
                visited[i] = false
            }
        }
    }

    dfs(0, 0)
    println(minDiff)
}

File: baekjoon_problems/exhaustive_search/14889.py
"""
두 팀의 능력치 차이가 최소가 되도록 만들어라

https://st-lab.tistory.com/122

브루트 포스을 통해 투 팀의 능력치 차이를 계산하여 최솟값을 갱신하는 방식으로 풀이했다. 조합을
사용하는 방법으로 문제를 풀이하면 된다. 블로그와 같이 visited를 사용해서 조금 더 나은 
시간 복잡도를 가지도록 풀이할 수 있다. 다만 중요한 것은 DFS를 사용해서 조합을 구하며 브루트
포스를 하는 것이 문제의 핵심 포인트이다.
"""

# 브루트 포스
def solution1():
    import sys
    import itertools

    # 참가자 수
    N = int(sys.stdin.readline().strip())
    # 능력치 
    graph = []

    for _ in range(N):
        graph.append(list(map(int, sys.stdin.readline().strip().split(" "))))
        
    min_diff = sys.maxsize

    # 스타트팀
    for start_team in list(itertools.combinations([i for i in range(1, N + 1)], N // 2)):
        link_team = []
        # 링크팀
        for i in range(1, N + 1):
            if i not in start_team:
                link_team.append(i)
        
        start_score = 0
        link_score = 0
        
        # 스타트팀 계산
        for x, y in list(itertools.combinations(start_team, 2)):
            start_score += graph[x - 1][y - 1]
            start_score += graph[y - 1][x - 1]
        
        # 링크팀 계산
        for x, y in list(itertools.combinations(link_team, 2)):
            link_score += graph[x - 1][y - 1]
            link_score += graph[y - 1][x - 1]
        
        # 두 팀의 능력치 차이 갱신
        min_diff = min(min_diff, abs(start_score - link_score))
        
    print(min_diff)